### PR TITLE
feat: Implement basic prepared statement support in IOx

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on https://github.com/influxdata/influxdb_iox/pull/6640~
re https://github.com/influxdata/influxdb_iox/issues/6471  and https://github.com/influxdata/influxdb_iox/issues/6482

Closes https://github.com/influxdata/influxdb_iox/pull/6495

This is enough to run a basic JDBC command line demo (aka https://github.com/influxdata/influxdb_iox/pull/6495)

While it supports prepared statements, it does not implement parameterized statements (I will implement that in phase 4).

# Changes
1. Implement `DoAction` endpoint handling
2. Add diagram with the prepared statements execution flow
3. Add a prepared statement API to our rust FlightSQL client (which I eventually plan to upstream)
4. Rust based end to end tests

# API notes
The client API uses the standard `prepare`/`execute` methods which is consistent with:
1. The  postgres / ANSI SQL  `PREPARE` / `EXECUTE` statements
2. The `tokio_postgres` model: https://docs.rs/tokio-postgres/0.7.7/tokio_postgres/struct.Client.html#method.execute
3. The existing names in  patterns set in arrow-rs (https://github.com/apache/arrow-rs/blob/de381ec544462597d60847f58a9e80dd11e1039c/arrow-flight/src/sql/client.rs#L159-L162)  from @avantgardnerio 

# Testing Note:
This PR only has a Rust based test. My next step is to add a JDBC integration test

Follow on steps:
1. Implement integration test (end to end using JDBC)
3. Implement parameter support (in phase 4)

### CODE REVIEW

File: service_grpc_flight/src/lib.rs [hunk 1]
```diff
@@ -271,9 +272,55 @@ type TonicStream<T> = Pin<Box<dyn Stream<Item = Result<T, tonic::Status>> + Send
 ///       ┃                                                  ┃
 /// ```
 ///
-/// ## FlightSQL Prepared Statement (NOT YET IMPLEMENTED)
+/// ## FlightSQL Prepared Statement (no bind parameters like $1, etc)
 ///
-/// TODO sequence diagram
+/// To run a prepared query, via FlightSQL, the client undertakes a
```
Comment (alamb) [inline service_grpc_flight/src/lib.rs:277]: here is a picture of the flow that was implemented

File: influxdb_iox/tests/end_to_end_cases/flightsql.rs [hunk 2]
```diff
@@ -62,6 +63,109 @@ async fn flightsql_query() {
...
+                    let connection = state.cluster().querier().querier_grpc_connection();
+                    let (channel, _headers) = connection.into_grpc_connection().into_parts();
+
+                    let mut client = FlightSqlClient::new(channel);
+
+                    // Add namespace to client headers until it is fully supported by FlightSQL
+                    let namespace = state.cluster().namespace();
+                    client.add_header("iox-namespace-name", namespace).unwrap();
+
+                    let handle = client.prepare(sql).await.unwrap();
```
Comment (alamb) [inline influxdb_iox/tests/end_to_end_cases/flightsql.rs:153]: here is an example of prepare/execute working end to end

File: influxdb_iox/tests/end_to_end_cases/cli.rs [hunk 3]
```diff
@@ -887,7 +887,7 @@ async fn query_ingester() {
                     // something like "wrong query protocol" or
                     // "invalid message" as the querier requires a
                     // different message format Ticket in the flight protocol
-                    let expected = "Unknown namespace: ";
+                    let expected = "Namespace '' not found";
```
Comment (alamb) [inline influxdb_iox/tests/end_to_end_cases/cli.rs:890]: this is due to cleaning up the error handling in the flight service

File: service_grpc_flight/src/lib.rs [hunk 4]
```diff
@@ -567,6 +655,23 @@ fn msg_from_descriptor(flight_descriptor: FlightDescriptor) -> Result<Any> {
     }
 }
 
+/// Figure out the namespace for this request, in this order:
```
Comment (alamb) [inline service_grpc_flight/src/lib.rs:658]: I plan to use this in my next PR to run the JDBC test end to end (as JDBC has no way currently to set the header name)

File: flightsql/src/planner.rs [hunk 5]
```diff
@@ -108,30 +129,123 @@ impl FlightSQLPlanner {
...
         }
     }
 }
 
-/// Decoded and validated FlightSQL command
+/// Represents a prepared statement "handle". IOx passes all state
+/// required to run the prepared statement back and forth to the
+/// client so any querier instance can run it
+#[derive(Debug, Clone)]
+struct PreparedStatementHandle {
```
Comment (tustvold) [inline flightsql/src/planner.rs:206]: I know it is a chore but this should probably be a protobuf so we have a forwards compatibility story. Otherwise it will be hard to rollout a change to this without it leading to errors

File: influxdb_iox_client/src/client/flightsql.rs [hunk 6]
```diff
@@ -156,4 +171,152 @@ impl FlightSqlClient {
...
+            return Err(FlightError::ProtocolError(format!(
+                "Expected 1 response for preparing a statement, got {}",
+                results.len()
+            )));
+        }
+        let result = results.pop().unwrap();
+
+        // decode the response
+        let response: arrow_flight::sql::Any = Message::decode(result.as_ref())
+            // TODO make this a real error upstream in FlightError
```
Comment (tustvold) [inline influxdb_iox_client/src/client/flightsql.rs:202]: ?

File: influxdb_iox_client/src/client/flightsql.rs [hunk 7]
```diff
@@ -156,4 +171,152 @@ impl FlightSqlClient {
...
+    } else {
+        Schema::try_from(IpcMessage(schema))?
+    };
+
+    Ok(Arc::new(schema))
+}
+
+/// try to unpack the [`arrow_flight::sql::Any`] as type `T`, returning Ok(None) if
+/// the type is wrong or Err if an error occurs
+fn try_unpack<T: ProstMessageExt>(msg: &arrow_flight::sql::Any) -> Result<Option<T>> {
```
Comment (tustvold) [inline influxdb_iox_client/src/client/flightsql.rs:265]: Do we need this? Can we use https://docs.rs/arrow-flight/latest/arrow_flight/sql/struct.Any.html#method.unpack instead?

File: influxdb_iox_client/src/client/flightsql.rs [hunk 8]
```diff
@@ -156,4 +171,152 @@ impl FlightSqlClient {
...
+    ) -> Self {
+        Self {
+            prepared_statement_handle,
+            dataset_schema,
+            parameter_schema,
+        }
+    }
+
+    /// returns the raw `prepared_statement_handle` from the server
+    pub fn into_handle(self) -> Bytes {
```
Comment (tustvold) [inline influxdb_iox_client/src/client/flightsql.rs:309]: This doesn't appear to be used?

File: service_grpc_flight/src/lib.rs [hunk 9]
```diff
@@ -521,16 +570,55 @@ where
...
+            .server
+            .db(&namespace_name, span_ctx.child_span("get namespace"))
+            .await
+            .context(NamespaceNotFoundSnafu {
+                namespace_name: &namespace_name,
+            })?;
+
+        let ctx = db.new_query_context(span_ctx);
+        let body = Planner::new(&ctx)
+            .flight_sql_do_action(&namespace_name, db, msg.clone())
```
Comment (tustvold) [inline service_grpc_flight/src/lib.rs:608]:
```suggestion
            .flight_sql_do_action(&namespace_name, db, msg)
```

File: flightsql/src/planner.rs [hunk 10]
```diff
@@ -108,30 +129,123 @@ impl FlightSQLPlanner {
...
+        }
+    }
+
+    /// Handles the action specified in `msg` and returns bytes for
+    /// the [`arrow_flight::Result`] (not the same as a rust
+    /// [`Result`]!)
+    pub async fn do_action(
+        namespace_name: impl Into<String>,
+        _database: Arc<dyn QueryNamespace>,
+        msg: Any,
```
Comment (tustvold) [inline flightsql/src/planner.rs:159]: Possibly something for a follow up, but I wonder if it might make for a cleaner API to not have `Any` in arguments, but instead in this case take `FlightSQLCommand` or something. The type-erasure is just a little hard to follow :sweat_smile: 

You even could lift the match into the parent, and have separate methods for each action / command on FlightSQLPlanner. :thinking:

File: flightsql/src/planner.rs [hunk 11]
```diff
@@ -108,30 +129,123 @@ impl FlightSQLPlanner {
...
+
+/// Decoded /  validated FlightSQL command messages
 #[derive(Debug, Clone)]
 enum FlightSQLCommand {
     CommandStatementQuery(String),
-    CommandPreparedStatementQuery(String),
+    /// Run a prepared statement
+    CommandPreparedStatementQuery(PreparedStatementHandle),
+    /// Create a prepared statement
+    ActionCreatePreparedStatementRequest(String),
```
Comment (tustvold) [inline flightsql/src/planner.rs:242]: I wonder if we should instead have a separate `FlightSQLAction` enumeration? Also see above comment about making the APIs more type-safe

File: service_grpc_flight/src/lib.rs [hunk 12]
```diff
@@ -38,6 +38,14 @@ use tracker::InstrumentedAsyncOwnedSemaphorePermit;
 /// for discussion on adding support to FlightSQL itself.
 const IOX_FLIGHT_SQL_NAMESPACE_HEADER: &str = "iox-namespace-name";
 
+/// Environment variable to take the FlightSQL name from
```
[the remark this answers is not in the document]
Comment (tustvold) [reply service_grpc_flight/src/lib.rs:46]: How are we intending for clients in general to interact with this functionality then?

File: service_grpc_flight/src/lib.rs [hunk 13]
```diff
@@ -266,9 +277,55 @@ type TonicStream<T> = Pin<Box<dyn Stream<Item = Result<T, tonic::Status>> + Send
...
+/// a prepared statement "handle".
+///
+/// 4. Encode the handle in a `CommandPreparedStatementQuery`
+/// FlightSQL structure in a [`FlightDescriptor`] and call the
+/// `GetFlightInfo` method with the the [`FlightDescriptor`]
 ///
-/// TODO sequence diagram
+/// 5. Steps 5,6,7 proceed the same as for a FlightSQL ad-hoc query
+///
+/// ```text
```
Comment (tustvold) [inline service_grpc_flight/src/lib.rs:299]: This is just... Sigh...

**Review by tustvold — APPROVED** [review]
Just some relatively minor nits

File: flightsql/src/planner.rs (hunk 5 above)
Comment (alamb) [reply flightsql/src/planner.rs:206]: That is a good point. I will file a ticket to track doing so

File: service_grpc_flight/src/lib.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (alamb) [reply service_grpc_flight/src/lib.rs:46]: My original belief was that there would be a proxy that added the appropriate namespace header (after doing authentication checks, etc). However @mhilton  has suggested this might not be necessary at all (see https://github.com/influxdata/influxdb_iox/pull/6693#discussion_r1086365274) -- I will clean it up in https://github.com/influxdata/influxdb_iox/pull/6693

File: flightsql/src/planner.rs (hunk 5 above)
Comment (alamb) [reply flightsql/src/planner.rs:206]: Filed https://github.com/influxdata/influxdb_iox/issues/6699

File: service_grpc_flight/src/lib.rs (hunk 13 above)
Comment (alamb) [reply service_grpc_flight/src/lib.rs:299]: just wait until you see it extended with what is needed for parameters 😭

File: influxdb_iox_client/src/client/flightsql.rs (hunk 6 above)
Comment (alamb) [reply influxdb_iox_client/src/client/flightsql.rs]: not sure -- removed

File: influxdb_iox_client/src/client/flightsql.rs (hunk 7 above)
Comment (alamb) [reply influxdb_iox_client/src/client/flightsql.rs]: 👍 Excellent call. Done in 97038c613

File: influxdb_iox_client/src/client/flightsql.rs (hunk 8 above)
Comment (alamb) [reply influxdb_iox_client/src/client/flightsql.rs:292]: It is not. I am (perhaps optimistically) dreaming of when I propose upstreaming all this code to arrow-flight in arrow-rs at which point I think it is important to be able to get this raw data. I will remove it for now.

File: service_grpc_flight/src/lib.rs (hunk 9 above)
Comment (alamb) [reply service_grpc_flight/src/lib.rs]: done

File: flightsql/src/planner.rs (hunk 10 above)
Comment (alamb) [reply flightsql/src/planner.rs:159]: My plan here is as follows: I will merge this PR as is, get https://github.com/influxdata/influxdb_iox/pull/6693 ready to go / tested and then create a new PR that refactors the structure to avoid the Any madness

File: flightsql/src/planner.rs (hunk 11 above)
Comment (alamb) [reply flightsql/src/planner.rs:242]: yes I think this is a good idea -- I will do it as a follow up per https://github.com/influxdata/influxdb_iox/pull/6667#discussion_r1086550005

File: service_grpc_flight/src/lib.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (alamb) [reply service_grpc_flight/src/lib.rs:46]: Update here is that this environment variable stuff is removed in https://github.com/influxdata/influxdb_iox/pull/6693 👍

File: flightsql/src/planner.rs (hunk 10 above)
Comment (alamb) [reply flightsql/src/planner.rs:159]: Refactor PRs 1: https://github.com/influxdata/influxdb_iox/pull/6703 and 2: https://github.com/influxdata/influxdb_iox/pull/6705/commits/a1a79a0070eb09ab17222f8ec8b08754725baeff
